dev/src/dev/db_tracking.clj

```diff
@@ -0,0 +1,124 @@
+(ns dev.db-tracking
+  "A set of utility function to track database changes.
+  Use this when you want to observe changes of database models when doing stuffs on UI.
+
+  How to use this?
+    > (track! models/Dashboard models/Card models/DashboardCard)
+    -- Go on UI and do stuffs like (i.e: update viz-settings of a dashcard).
+
+    > (changes)
+    ;; => {:report_card {:insert ...}}
+
+  You can use [[reset-changes!]] to clear our all the current trackings.
+  And [[untrack-all!]] or [[untrack!]] to stop tracking."
+  (:require
+   [clojure.pprint :as pprint]
+   [metabase.util :as u]
+   [methodical.core :as m]
+   [toucan2.core :as t2]
+   [toucan2.model :as t2.model]
+   [toucan2.tools.before-delete :as t2.before-delete]
+   [toucan2.tools.before-insert :as t2.before-insert]
+   [toucan2.tools.before-update :as t2.before-update]))
+
+(def ^:private tracking (atom {}))
+
+(def ^:private tracked-models (atom #{}))
+
+(defn- clean-tracking
+  [tracking]
+  (dissoc tracking :updated_at :created_at))
+
+(defn- new-tracking
+  "Add a tracking to the [[tracking]] atom.
+
+    > (new-tracking :models/Card :insert {:name \"new card\"})
+    instance
+
+    > @tracking
+    {:insert {:report_card [{:name \"new card\"}]}]}.
+
+  For insert, track the instance as a map.
+  For update, only track the changes."
+  [model action row-or-instance]
+  (let [model      (t2/resolve-model model)
+        track-item (->> (case action
+                          :update
+                          (into {} (t2/changes row-or-instance))
+                          (into {} row-or-instance))
+                       clean-tracking)
+        path       [(t2/table-name model) action]]
+    ;; ideally this should be debug, but for some reasons this doesn't get logged
+    (println (u/colorize :magenta :new-tracking) (u/colorize :magenta path))
+    (pprint/pprint track-item)
```

Comment:
I'd prefer if it tapped this data to a tap target, so I could customize how it gets printed. It requires extra customization from the developer though. If you extracted this to a function I could redef it myself, to do something like this:
```suggestion
    (tap> [:tracking-change path track-item])
```